src/librustc_resolve/lib.rs

```diff
@@ -99,16 +99,56 @@ use std::usize;
 use resolve_imports::{Target, ImportDirective, ImportResolution};
 use resolve_imports::Shadowable;
 
-
 // NB: This module needs to be declared first so diagnostics are
 // registered before they are used.
 pub mod diagnostics;
 
+macro_rules! resolve_err {
+    ($this:expr, $span:expr, $code:ident, $($rest:tt)*) => {
+        if $this.emit_errors {
+            span_err!($this.session, $span, $code, $($rest)*);
+        }
+    }
+}
+
 mod check_unused;
 mod record_exports;
 mod build_reduced_graph;
 mod resolve_imports;
 
+pub enum ResolutionError<'b, 'a:'b, 'tcx:'a> {
+    /// error: static variables cannot be referenced in a pattern
+    StaticVariableReference(&'b Resolver<'a, 'tcx>, syntax::codemap::Span),
+    /// error: does not name a struct
+    DoesNotNameAStruct(&'b Resolver<'a, 'tcx>, syntax::codemap::Span),
+    /// error: is a struct variant name, but this expression uses it like a function name
+    StructVariantUsedAsFunction(&'a Resolver<'a, 'tcx>, syntax::codemap::Span),
+    /// error: unresolved import
+    UnresolvedImport(&'b Resolver<'a, 'tcx>, syntax::codemap::Span),
+    /// error: failed to resolve
+    FailedToResolve(&'b Resolver<'a, 'tcx>, syntax::codemap::Span),
+}
+
+fn resolve_error<'b, 'a:'b, 'tcx:'a>(resolution_error: &ResolutionError<'b, 'a, 'tcx>, formatted: &str) {
+    match resolution_error {
```

Comment:
Instead of passing in `formatted` from the outside (and thus requiring clients to construct the appropriate message for each variant of `ResolutionError`, I would prefer a structure where the dispatching here embeds the appropriate error message for the `ResolutionError` variant.

Doing this well may require attaching extra state to one or more of the variants of `ResolutionError`.  (And that is okay, it is in fact good practice, IMO.)